### PR TITLE
[3D-parallelism] Hybrid Model Parallelism

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
New features

### PR changes
APIs

### Describe
* new features
  * Hybrid Model Parallelism： 
       * Combine 3 single Model parallelism strategies (megatron, sharding, pipline) together as a Hybrid Parallelism strategy
       * Uniform switch to turn on/off each single parallelism strategy (temporally use Sharding & Pipeline as the uniform API)
* performance optimization：
    * speed
        *  remove potential un-necessary sync_calc & sync_calc in hybrid model parallelism
    * Memory usage
        * optimizer offload
        * optimizer temporary vars inplacement  [commit withdraw, update in next pr]

* **performance-related**
    *  the order of parallelism from inner to outer is : mp --> sharidng --> pp
    *  mp (megatron) and sharding parallelism will introduce large communication and is recommend: **to be used within node. (mp_degree * sharding_degree = number of gpu per node)**
    * pp parallelism have less communication load compared with above two, which make them **more suitable to use across the nodes  (pp_degree = number node)**

example
* assume we have 4 nodes with 8 gpus per node:

*  mp-sharding-pp 3D parallelism
```python
        dist_strategy.sharding = True
        dist_strategy.pipeline = True
        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
                                            "pp_degree": 4,
                                            "sharding_degree":4,
                                            "mp_degree": 2,
                                            "optimize_offload": True,
                                            }
        dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
                                            "micro_batch_size": 1,
                                            "accumulate_steps": 4,
                                            }

```

*  mp-pp 2D parallelism
```python
        dist_strategy.sharding = True
        dist_strategy.pipeline = True
        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
                                            "pp_degree": 4,
                                            "sharding_degree":1,
                                            "mp_degree": 8,
                                            "optimize_offload": True,
                                            }
        dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
                                            "micro_batch_size": 1,
                                            "accumulate_steps": 4,
                                            }
```

* shardign-pp 2D parallelism
```python
        dist_strategy.sharding = True
        dist_strategy.pipeline = True
        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
                                            "pp_degree": 1,
                                            "sharding_degree":4,
                                            "mp_degree": 8,
                                            "optimize_offload": True,
                                            }
        dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
                                            "micro_batch_size": 1,
                                            "accumulate_steps": 4,
                                            }

```

*  mp-sharding 2D parallelism
```python
        dist_strategy.sharding = True
        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
                                            "pp_degree": 1,
                                            "sharding_degree":4,
                                            "mp_degree": 8,
                                            "optimize_offload": False,
                                            "gradient_merge_acc_step": 4,
                                            }

```

### CODE REVIEW

File: paddle/fluid/framework/distributed_strategy.proto [hunk 1]
```diff
@@ -29,14 +29,18 @@ message RecomputeConfig {
 }
 
 message ShardingConfig {
-  optional float segment_broadcast_MB = 1 [ default = 32.0 ];
-  optional bool hybrid_dp = 2 [ default = false ];
-  optional int32 sharding_degree = 3 [ default = 8 ];
-  optional int32 mp_degree = 4 [ default = 1 ];
-  optional string sharding_segment_strategy = 5
+  optional string sharding_segment_strategy = 1
```
Comment (wangxicoding) [inline paddle/fluid/framework/distributed_strategy.proto:32]: Enum comments

File: paddle/fluid/framework/distributed_strategy.proto [hunk 2]
```diff
@@ -29,14 +29,18 @@ message RecomputeConfig {
...
-  optional int32 gradient_merge_acc_step = 7 [ default = 1 ];
+  optional float segment_broadcast_MB = 2 [ default = 32.0 ];
+  repeated string segment_anchors = 3;
+  optional int32 sharding_degree = 4 [ default = 8 ];
+  optional int32 mp_degree = 5 [ default = 1 ];
+  optional int32 dp_degree = 6 [ default = 1 ];
+  optional bool hybrid_dp = 7 [ default = false ];
+  optional int32 gradient_merge_acc_step = 8 [ default = 1 ];
+  optional bool optimize_offload = 9 [ default = false ];
+  optional bool pp_allreduce_in_optimize = 10 [ default = false ];
```
Comment (wangxicoding) [inline paddle/fluid/framework/distributed_strategy.proto:42]: Add some comments, in 3d or 4d parallel, allreduce_in_optimize=True can reduce communication,  allreduce_in_optimize=False can reduce memory

File: paddle/fluid/framework/distributed_strategy.proto (hunk 1 above)
Comment (JZ-LIANG) [reply paddle/fluid/framework/distributed_strategy.proto:32]: recorded,  document will be added in fluiddoc and fleetx

File: paddle/fluid/framework/distributed_strategy.proto (hunk 2 above)
Comment (JZ-LIANG) [reply paddle/fluid/framework/distributed_strategy.proto:42]: recorded, document will be added in fluiddoc and fleetx and .py file where the feature is called.

but I think this should be a feature for internal project now, and we should not  expose  It to users ?

**Review by fuyinno4 — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM for backward.py

**Review by wangxicoding — APPROVED** [review]
LGTM

File: paddle/fluid/framework/distributed_strategy.proto (hunk 1 above)
Comment (wangxicoding) [reply paddle/fluid/framework/distributed_strategy.proto:32]: Also need add comments to this code.

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**